File: src/main/java/me/ztowne13/customcrates/listeners/BlockBreakListener.java
package me.ztowne13.customcrates.listeners;

import me.ztowne13.customcrates.CustomCrates;
import me.ztowne13.customcrates.Messages;
import me.ztowne13.customcrates.SettingsValues;
import me.ztowne13.customcrates.crates.Crate;
import me.ztowne13.customcrates.crates.PlacedCrate;
import me.ztowne13.customcrates.crates.options.ObtainType;
import me.ztowne13.customcrates.players.PlayerManager;
import me.ztowne13.customcrates.utils.CrateUtils;
import org.bukkit.GameMode;
import org.bukkit.Location;
import org.bukkit.entity.Player;
import org.bukkit.event.EventHandler;
import org.bukkit.event.Listener;
import org.bukkit.event.block.BlockBreakEvent;

public class BlockBreakListener implements Listener
{
    CustomCrates cc;

    public BlockBreakListener(CustomCrates cc)
    {
        this.cc = cc;
    }

    @EventHandler
    public void onBreakPlacedCrate(BlockBreakEvent e)
    {
        Player p = e.getPlayer();
        Location l = e.getBlock().getLocation();

        if (PlacedCrate.crateExistsAt(cc, l))
        {
            PlacedCrate cm = PlacedCrate.get(cc, l);
            Crate crates = cm.getCrates();

            if (crates.getCs().getOt().isStatic())
            {
                if (!p.hasPermission("customcrates.deletestatic"))
                {
                    e.setCancelled(true);
                    Messages.FAILED_BREAK_CRATE.msgSpecified(cc, p, new String[]{"%crate%", "%reason%"},
                            new String[]{crates.getName(), "static"});
                }
                return;
            }

            cm.delete();
            Messages.BROKEN_CRATE.msgSpecified(cc, p, new String[]{"%crate%"}, new String[]{crates.getName()});
        }
    }

    @EventHandler
    public void onBreakForLuckyChest(BlockBreakEvent e)
    {
        Player p = e.getPlayer();
        Location l = e.getBlock().getLocation();

        // Crate doesn't exist at location
        if(!PlacedCrate.crateExistsAt(cc, l))
        {
            // Event isn't already cancelled
            if (!e.isCancelled())
            {
                // Not in creative mode or creative mode is allowed
                if (!p.getGameMode().equals(GameMode.CREATIVE) || (Boolean) SettingsValues.LUCKYCHEST_CREATIVE.getValue(cc))
                {
                    // Luckychests enabled
                    if (PlayerManager.get(cc, p).getPdm().isActivatedLuckyChests())
                    {
                        // Cycle through all potential crates
                        for (Crate crates : Crate.getLoadedCrates().values())
                        {
                            // Check if the crate is a lucky chest and if it is enabled
                            if (CrateUtils.isCrateUsable(crates) && crates.getCs().clcExists() &&
                                    crates.getCs().getOt().equals(ObtainType.LUCKYCHEST))
                            {
                                // Check if the lucky chesty should be placed at the location
                                if (crates.getCs().getClc().checkRun(e.getBlock()))
                                {
                                    if((!e.getBlock().hasMetadata("PLACED") || e.getBlock().getMetadata("PLACED") == null) || (Boolean) SettingsValues.LUCKYCHEST_ALLOW_PLACED_BLOCKS.getValue(cc))
                                    {
                                        PlacedCrate cm = PlacedCrate.get(cc, e.getBlock().getLocation());
                                        cm.setup(crates, true);
                                        Messages.FOUND_LUCKY_CHEST.msgSpecified(cc, p);
                                        e.setCancelled(true);
                                        break;
                                    }
                                }
                            }
                        }
                    }
                }
            }
        }
    }
}

File: src/main/kotlin/me/ztowne13/customcrates/listeners/BlockBreakListener.kt
package me.ztowne13.customcrates.listeners

import org.bukkit.GameMode
import org.bukkit.event.block.BlockBreakEvent
import me.ztowne13.customcrates.SpecialisedCrates
import me.ztowne13.customcrates.crates.Crate
import me.ztowne13.customcrates.crates.PlacedCrate
import me.ztowne13.customcrates.crates.options.ObtainType
import me.ztowne13.customcrates.players.PlayerManager
import org.bukkit.event.EventHandler
import org.bukkit.event.Listener

class BlockBreakListener(var cc: SpecialisedCrates) : Listener {
    @EventHandler
    fun onBreakPlacedCrate(e: BlockBreakEvent) {
        val p = e.player
        val l = e.block.location
        if (PlacedCrate.Companion.crateExistsAt(l)) {
            val cm: PlacedCrate = PlacedCrate.Companion.get(cc, l)
            val crates = cm.crate
            if (crates!!.settings.obtainType!!.isStatic) {
                if (!p.hasPermission("me.ztowne13.customcrates.admin") && !p.hasPermission("specializedcrates.admin")) {
                    e.isCancelled = true
                    Messages.FAILED_BREAK_CRATE.msgSpecified(
                        cc, p, arrayOf("%crate%", "%reason%"), arrayOf(
                            crates.displayName, "static"
                        )
                    )
                }
                return
            }
            cm.delete()
            Messages.BROKEN_CRATE.msgSpecified(
                cc, p, arrayOf("%crate%"), arrayOf(
                    crates.displayName
                )
            )
        } else {
            // Event isn't already cancelled
            if (!e.isCancelled) {
                // Not in creative mode or creative mode is allowed
                if (p.gameMode != GameMode.CREATIVE || (SettingsValue.LUCKYCHEST_CREATIVE.getValue(cc) as Boolean)) {
                    // Luckychests enabled
                    if (PlayerManager.Companion.get(cc, p)!!.getPlayerDataManager().isActivatedLuckyChests()) {
                        // Cycle through all potential crates
                        for (crates in Crate.Companion.getLoadedCrates().values) {
                            // Check if the crate is a lucky chest and if it is enabled
                            if (CrateUtils.isCrateUsable(crates) && crates.settings.luckyChestSettingsExists() && crates.settings.obtainType == ObtainType.LUCKYCHEST) {
                                // Check if the lucky chesty should be placed at the location
                                if (crates.settings.luckyChestSettings!!.checkRun(e.block)) {
                                    // Check if this block is a placed block or not and whether or not that's okay
                                    if (!e.block.hasMetadata("PLACED") ||
                                        e.block.getMetadata("PLACED") == null ||
                                        (SettingsValue.LUCKYCHEST_ALLOW_PLACED_BLOCKS.getValue(cc) as Boolean)
                                    ) {
                                        // Check to make sure the player has the permission or doesn't need the permission
                                        if (!crates.settings.luckyChestSettings!!.isRequirePermission ||
                                            e.player.hasPermission(crates.settings.permission!!)
                                        ) {
                                            val cm: PlacedCrate = PlacedCrate.Companion.get(
                                                cc, e.block.location
                                            )
                                            cm.setup(crates, true)
                                            Messages.FOUND_LUCKY_CHEST.msgSpecified(cc, p)
                                            e.isCancelled = true
                                            break
                                        }
                                    }
                                }
                            }
                        }
                    }
                }
            }
        }
    }
}